### PR TITLE
rgw: s3select: support SQL IN, LIKE, IS NULL

### PR DESCRIPTION
**Pull request by albin-antony**

### CODE REVIEW

File: include/s3select_functions.h [hunk 1]
```diff
@@ -867,6 +869,35 @@ struct _fn_isnull : public base_function
...
+    int args_size = args->size();
+    while (args_size > 1)
+    {
+      base_statement* expr = *iter;
+      value expr_val = expr->eval();
+      iter++;
+      if ( expr_val == main_expr_val) {
+        result->set_value((int64_t)true);
+      } else {
+        result->set_value((int64_t)false);
```
Comment (galsalomon66) [inline include/s3select_functions.h:895]: /in/ should return false only upon *all* expressions are not equal.

File: include/s3select_functions.h [hunk 2]
```diff
@@ -85,6 +86,7 @@ class s3select_functions
     {"nullif", s3select_func_En_t::NULLIF},
     {"between", s3select_func_En_t::BETWEEN},
     {"isnull", s3select_func_En_t::IS_NULL},
+    {"in", s3select_func_En_t::IN},
```
Comment (galsalomon66) [inline include/s3select_functions.h:91]: review 
/void push_in_predicate::operator()/
what is the operator name?

Comment (albin-antony) [reply include/s3select_functions.h:91]: Thanks. Will update that

File: include/s3select_functions.h [hunk 3]
```diff
@@ -867,6 +869,35 @@ struct _fn_isnull : public base_function
...
+    bs_stmt_vec_t::iterator it = args->end();
+    base_statement* main_expr = *it;
+    value main_expr_val = main_expr->eval();
+    int args_size = args->size();
+    while (args_size > 1)
+    {
+      base_statement* expr = *iter;
+      value expr_val = expr->eval();
+      iter++;
+      if ( expr_val == main_expr_val) {
```
Comment (galsalomon66) [inline include/s3select_functions.h:892]: what happens upon comparing expression not of the same type (string , int ) ?

**Comment by galsalomon66** [comment]
the /like/ operator has some optimizations issues to resolve
upon query such as _' expression like "%aa%" '_ the right expression is "constant" per query lifetime, using this fact will improve performance.
the boost regular-expression is not as sql-like regular-expression, thus a translation to sql-like should be done.

File: include/s3select_functions.h [hunk 4]
```diff
@@ -846,6 +855,108 @@ struct _fn_version : public base_function
...
+    while (args_size > 1)
+    {
+      base_statement* expr = *iter;
+      value expr_val = expr->eval();
+      iter++;
+      if ( expr_val.type == main_expr_val.type) {
+        if ( expr_val == main_expr_val)
+      {
+        result->set_value((int64_t)true);
+        flag = 1;
```
Comment (galsalomon66) [inline include/s3select_functions.h:899]: it keep scan after a match expression, it should stop and set result=true

File: include/s3select_functions.h [hunk 5]
```diff
@@ -846,6 +855,112 @@ struct _fn_version : public base_function
...
+    {
+      base_statement* expr = *iter;
+      value expr_val = expr->eval();
+      iter++;
+      if ( expr_val.type == main_expr_val.type) {
+        if ( expr_val == main_expr_val) {
+          result->set_value((int64_t)true);
+          break;
+        } else {
+          result->set_value((int64_t)false);
```
Comment (galsalomon66) [inline include/s3select_functions.h:899]: it means it return false for the first expression which return false.

/in/ should return false , *only* upon all expression are false. 
and return true for any expression which return true (at least one expression)

File: include/s3select_functions.h [hunk 6]
```diff
@@ -846,6 +855,119 @@ struct _fn_version : public base_function
...
+  {
+    bs_stmt_vec_t::iterator iter = args->begin();
+    base_statement* expr = *iter;
+    iter++;
+    base_statement* main_expr = *iter;
+    value expr_val = expr->eval();
+    value main_expr_val = main_expr->eval();
+    std::string a = expr_val.to_string();
+    std::string s = transform(a);
+    std::regex x{s};
```
Comment (galsalomon66) [inline include/s3select_functions.h:961]: the regex is actually compiled expression, which _regex_match_ is using.
/transform/ and /regex/ should called once per query ,from performance perspective it's a big difference.
thus, the optimized way is to set /x/ once on construction (should add construction).

/push_like_predicate/ is the place to call constructor with the like expression.
the parser definition should modified also , to accept constant string and not to accept general expression.

**Comment by galsalomon66** [comment]
what about unit-test for /in/ statements ?

File: include/s3select_functions.h [hunk 7]
```diff
@@ -846,6 +855,119 @@ struct _fn_version : public base_function
...
+};
+
+struct _fn_in : public base_function
+{
+
+  value res;
+  
+  bool operator()(bs_stmt_vec_t* args, variable* result)
+  {
+    bs_stmt_vec_t::iterator iter = args->begin();
```
Comment (mattbenjamin) [inline include/s3select_functions.h:884]: please use meaningful and consistent variable names;  if you really need temporary references to begin() and end(), use something like iter_begin and iter_end.  "iter" and "it" are both conventions for an iterator where there can be no ambiguity about what it points to, such as the iterator in a loop over sequence;  but they are -different conventions-;  you don't use "it" and "iter" together;  I prefer iter--so dos Lakos :)

File: include/s3select_functions.h [hunk 8]
```diff
@@ -846,6 +855,119 @@ struct _fn_version : public base_function
...
+};
+
+struct _fn_like : public base_function
+{
+
+  value res;
+
+  std::string transform(std::string s)
+  {
+    bool startswith = (s[0] == '%' && s[s.size()-1] != '%');
```
Comment (galsalomon66) [inline include/s3select_functions.h:914]: looking good. 
i had tried several combinations, and results were OK per those runs.

there are endless combinations, we should seek for a good coverage for such an operator.

File: include/s3select_functions.h [hunk 9]
```diff
@@ -790,6 +801,47 @@ struct _fn_utcnow : public base_function
...
+    base_statement* main_expr = *iter;
+
+    //TODO types validations 
+    value second_expr_val = second_expr->eval();
+    value first_expr_val = first_expr->eval();
+    value main_expr_val = main_expr->eval();
+
+    if((main_expr_val > first_expr_val) && (main_expr_val < second_expr_val))
+    {
+      result->set_value((int64_t)true);
```
Comment (galsalomon66) [inline include/s3select_functions.h:834]: we can avoid casting , by adding boolean to /set_value/ interface.

Comment (albin-antony) [reply include/s3select_functions.h:834]: ok, will add that

File: test/s3select_test.cpp [hunk 10]
```diff
@@ -905,6 +955,405 @@ TEST(TestS3selectFunctions, nulladdition)
...
+        false, // dont skip last line
+        true   // aggregate call
+        ); 
+    ASSERT_EQ(status, 0); 
+    ASSERT_EQ(s3select_result, std::string("true,\n"));
+}
+
+TEST(TestS3selectFunctions, inoperatorfalse)
+{
+    s3select s3select_syntax;
```
Comment (galsalomon66) [inline test/s3select_test.cpp:1340]: add unit test with mix types in set , and complex expression. 

similar to 
{ int(_1)*int(_2)  in  ( 111 ,111.0 ,"111" , (10+11) )  }

File: include/s3select_functions.h [hunk 11]
```diff
@@ -956,13 +963,13 @@ struct _fn_like : public base_function
     base_statement* main_expr = *iter;
     value expr_val = expr->eval();
     value main_expr_val = main_expr->eval();
-    std::string a = expr_val.to_string();
-    std::string s = transform(a);
-    std::regex x{s};
+    std::string stringvalue = expr_val.to_string();
+    std::string stringtransform = transform(stringvalue);
+    std::regex x{stringtransform};
```
Comment (galsalomon66) [inline include/s3select_functions.h:968]: please **note**:
/_std::regex x_/ is defined twice , here , inside /_operator()_/
and also as /__fn_like_/ data member 

it should defined once, and populate once, i.e. transform should run upon construction.

File: include/s3select_functions.h [hunk 12]
```diff
@@ -908,14 +908,21 @@ struct _fn_like : public base_function
 {
 
   value res;
+  std::regex x;
```
Comment (galsalomon66) [inline include/s3select_functions.h:911]: please use better name (something like compiled-regex)
review comment bellow.

constructor should run transform on right expression (should be constant string in s3select query)
and populate x (actually constructing x)

File: test/s3select_test.cpp [hunk 13]
```diff
@@ -1352,7 +1392,27 @@ TEST(TestS3selectFunctions, inoperatorfalse)
...
     ASSERT_EQ(status, 0); 
-    ASSERT_NE(s3select_result, std::string("true,\n"));
+    ASSERT_EQ(s3select_result, std::string("true,\n"));
+}
+
+TEST(TestS3selectFunctions, inoperatormore)
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select \"true\" from stdin where \"a\" in (\"b\", \"a\", \"d\", \"e\", \"f\");" ;
+    auto status = s3select_syntax.parse_query(input_query.c_str());
```
Comment (galsalomon66) [inline test/s3select_test.cpp:1402]: goot test for /IN/
please add the mix types (and not just constant , also use expressions) , to make sure operator can handle that.

something like (handle different type comparison) 
where 10 in ( 5.0*2.0 , 12+1 , 9+1.2)

**Comment by galsalomon66** [comment]
"aabbccdef" like "%de_";'  return false
 "aabbccdef" like "_ab%";'  return false

**Comment by galsalomon66** [comment]
good coverage for like operator.
a combination of predicates  (/in/ , /like/ , /is-null/ , /between/ ) in single where clause, will be a good unit-test.

File: include/s3select_functions.h [hunk 14]
```diff
@@ -965,13 +973,10 @@ struct _fn_like : public base_function
     base_statement* main_expr = *iter;
     value expr_val = expr->eval();
```
Comment (galsalomon66) [inline include/s3select_functions.h:974]: there is no need for expr , its not used
expr->eval() is a waste of CPU time

Comment (albin-antony) [reply include/s3select_functions.h]: correct. Updated

File: include/s3select_functions.h [hunk 15]
```diff
@@ -803,6 +853,131 @@ struct _fn_version : public base_function
...
+        pos += replaceWith.length();
+      }
+      std::replace(s.begin(), s.end(), '_', '.');
+    }
+  }
+  
+  bool operator()(bs_stmt_vec_t* args, variable* result)
+  {
+    bs_stmt_vec_t::iterator iter = args->begin();
+    base_statement* expr = *iter;
```
Comment (galsalomon66) [inline include/s3select_functions.h:968]: expr is not used, need to remove it.

**Comment by galsalomon66** [comment]
in /_fn_between/
need to check type 
a between left and right 

a,left,right should belong to same type (string,numeric)

**Comment by albin-antony** [comment]
> in /_fn_between/
> need to check type
> **a** between **left** and **right**
> 
> a,left,right should belong to same type (string,numeric)

updated

**Pull request merged**